File: 프로그래머스/3/43162. 네트워크/네트워크.cpp
#include <bits/stdc++.h>
using namespace std;
typedef vector<vector<int>> vi;
const int N = 300;
int p[N];

int find(int x)
{
    if (p[x] == x) return x;
    else return p[x] = find(p[x]);
}

int solution(int n, vi a) 
{
    for (int i = 1; i <= n; i++) p[i] = i;
    for (int i = 0; i < n; i++) for (int j = 0; j < n; j++)
    {
        if (a[i][j] == 1)
        {
            int u = find(i + 1), v = find(j + 1);
            p[u] = v;
        }
    }
    
    int ans = 0;
    
    for (int i = 1; i <= n; i++)
    {
        if (p[i] == i)
            ans++;
    }
    
    return ans;
}

File: 프로그래머스/3/43162. 네트워크/네트워크.java
class Solution 
{
    public static final int N = 300;
    int[] p = new int[N];
    int n;
    
    public int find(int x)
    {
        if (p[x] == x) return x;
        else return p[x] = find(p[x]);
    }
    
    public int solution(int N, int[][] edge) 
    {
        n = N;
        for (int i = 1; i <= n; i++) p[i] = i;
        
        int sz = edge.length;
        
        for (int i = 0; i < sz; i++) for (int j = 0; j < sz; j++)
        {
            if (edge[i][j] == 1)
            {
                int u = i + 1, v = j + 1;
                u = find(u); v = find(v);
                if (u == v) continue;
                p[u] = v;
            }
        }
        
        int ans = 0;
        
        for (int i = 1; i <= n; i++)
        {
            if (p[i] == i)
                ans++;
        }
        
        return ans;
    }
}